### ISSUE TITLE
Consider switching from gulp to webpack

### ISSUE THREAD

**Issue by danielballan** [edgi-govdata-archiving/web-monitoring-ui#88]
The archivers' repos are using webpack, so switching would reduce the number of different technologies in the EDGI stack. Personally, I have more experience debugging webpack than gulp, but I think @lightandluck and @Mr0grog's opinions are more relevant.

**Comment by Mr0grog** [comment]
I would say Webpack is more of a replacement for Browserify than Gulp—it can do a little bit of what Gulp does but is in many ways orthogonal; a lot of projects use both Webpack and Gulp.

That said, we aren't using Gulp to do many of the things we could/should with it (really, if we’re using Gulp, we shouldn’t be calling anything directly with NPM scripts, so we should be managing app running, linting, tests, etc.), so getting rid of Browserify would remove everything we currently use Gulp for :P

Anyway! I am somewhat agnostic about this. Generally, I like smaller, less configuration-y tools like Browserify where you explicitly layer on the functionality you want, but both Webpack and Browserify are well built and each have their own pitfalls in my experience. Browserify does the most simple thing really easily, but you have to think about how to best approach everything after that for your particular project. Webpack gives you a fairly well defined, happy path do most common things that go beyond the ultra-simple without having to think too much at all (I often compare them to Flask vs. Django or Sinatra vs. Rails).

**TL;DR** Most of the time I have a slight preference toward Browserify, but I think the shared technology argument is a stronger one here. I’d be perfectly happy to switch to Webpack and also drop Gulp (since we aren’t using it much anyhow). I think it makes sense to do this at the same time as #53, since that will require re-tooling our build process anyway.

**Comment by lightandluck** [comment]
I don't have much experience with either so have no personal preference. 
Totally makes sense combining tech stacks.
![Powers combine](http://www.themorejewknow.com/uploads/1/2/0/9/12094351/6826869_orig.gif)

**Comment by b5** [comment]
Moving this comment from #92 for UI library & transpiler discussion:

#### Regarding Webpack 
> One suggestion, moving to webpack 2 as the transpiler will give you tree-shaking, which really shines when you start pulling in these bloated UI libraries. It'll mean that you can import with impunity, knowing that the production build will eliminate dead code for you.

#### React Component Libraries 

From @lightandluck:
> On the subject, would it make sense to use the larger component libraries? These all look very active: [React-Bootstrap](https://github.com/react-bootstrap), [Material UI](https://github.com/callemall/material-ui), [Pivotal UI](https://github.com/pivotal-cf/pivotal-ui). I've thrown out the idea of using tabs for workflow in future iterations. So it might make sense to use a library that we can just pull different UI components from. Pivotal seems ideal, because you don't have to import every component, but can import individual ones and a small core library to run it.

@b5 response:
> Looks like material-ui uses typescript, and is structured with the intent that you won't wander far outside of it's component library. I'd knock that one off the list straight out the gates. Also, styles are inlined into the js, which are subject of rigorous debate.
> Pivotal-UI & React-Bootstrap both look like great options. After browsing through their components, they're both great react projects IMHO. I'd say drawing from either would be lovely, I'd go with bootstrap if only because I find it less ugly.

**Comment by Mr0grog** [comment]
Re: webpack, I think we're all pretty on-board with making that switch. I got halfway into it with #53, but felt like it was too big in scope to see what was changing and clearly evaluate. I’ve failed to get back to it, so this is really just me dropping the ball.

Component libs: I’ll defer to your deeper analysis on material-ui, @b5. Let's strike that from consideration.

As for the others, I'm not especially keen on adding these kinds of tools that are *mostly* more focused on layout than interaction/functionality—my experience is that you often wind up going through a lot of effort to override them all over the place as you develop your site/app’s aesthetic. I don't think switching from Bootstrap to React-Bootstrap would solve problems like [this one](https://github.com/edgi-govdata-archiving/web-monitoring-ui/pull/92#discussion_r132536437), for example. Open to other arguments on that, though.

**Comment by b5** [comment]
I'd be happy to jump in on the webpack stuff, as I've had to suffer through config more than I'd like to admit. Should be able to get a PR opened up this week for discussion.

I'm tempted to agree with you on the component library approach @Mr0grog. I find batteries-included solutions to be great right up until the moment they don't provide an obvious way to execute some special functionality you're looking to accomplish, and then frustration sets in as you're forced to go against the grain of the lib. Conversely, if your needs fit inside of the bounds of the library, productivity gains are pretty huge.

This gives food for thought in terms of creating our own lib out of the components we already have. To me a library that _doesn't_ inline default styles, and instead focuses on scaffolding types of interactions, with styling included as importable css would be worth pursuing.

**Comment by Mr0grog** [comment]
> I'd be happy to jump in on the webpack stuff

That’d be great!

> To me a library that doesn't inline default styles, and instead focuses on scaffolding types of interactions, with styling included as importable css would be worth pursuing.

👍 👍 That’s the flavor of UI lib I’ve always found most useful, too. Will try and comment on that context issue today.

**Comment by lightandluck** [comment]
Wanted to leave some notes for things I found that would be useful when switching to webpack.

Currently, we use `babel-preset-es2015` and `babel-preset-es2016` packages. We can switch to [`babel-preset-env`](https://babeljs.io/docs/plugins/preset-env/), which supersedes those and uses  [browserlist](https://github.com/ai/browserslist) which is a nifty way to set which browsers to target for compilation. Being able to do the below is awesome:
```
'targets': {
      'browsers': ['last 2 versions']
}
```

Also, thought including `postcss-loader` with `autoprefixer` would be very useful. They also use the same browserlist syntax to target browsers.

**Comment by lightandluck** [comment]
One thing I noticed is that `browserlist` should be added to `package.json` and everything should work with it. In my limited testing though, `autoprefixer` worked with the setting in `package.json` but `babel-preset-env` did not. `babel-preset-env` is fairly new so only the beta version worked with that setting. The current stable version has to include the setting in `.babelrc` or inline.
